### PR TITLE
Evaleev/feature/template task graph

### PR DESCRIPTION
**Pull request by evaleev**

### CODE REVIEW

File: ttg/ttg/base/tt.h [hunk 1]
```diff
@@ -100,29 +99,36 @@ namespace ttg {
...
 
   private:
+   // non-copyable
    TTBase(const TTBase &) = delete;
    TTBase &operator=(const TTBase &) = delete;
-   TTBase(TTBase &&) = delete;
-   TTBase &operator=(TTBase &&) = delete;
+
+   // movable
+   TTBase(TTBase && other) : instance_id(other.instance_id)
```
Comment (devreal) [inline ttg/ttg/base/tt.h:107]: Is it intentional to make the move ctor/op private? Does that actually allow moving?

File: ttg/ttg/base/tt.h [hunk 2]
```diff
@@ -20,21 +20,20 @@ namespace ttg {
     }
   } // namespace detail
 
-  /// Provides basic information and graph connectivity (eventually statistics,
-  /// etc.)
+  /// A base class for all template tasks
   class TTBase {
   private:
-    uint64_t instance_id;  //< Unique ID for object
+    int64_t instance_id;  //!< Unique ID for object; in after-move state will be -1
```
Comment (devreal) [inline ttg/ttg/base/tt.h:26]:
```suggestion
    int64_t instance_id;  //!< Unique ID for object; -1 signals invalid state
```

File: ttg/ttg/tt.h [hunk 3]
```diff
@@ -11,50 +11,65 @@
 
 namespace ttg {
 
+  /// @brief a template task graph implementation
+
```
Comment (devreal) [inline ttg/ttg/tt.h:15]: Can you add a sentence on how a TT is different from a TTG? The CompositeOp kind of made clear that it is a composite of Ops but I'm not sure that's clear anymore.

File: ttg/ttg/tt.h [hunk 4]
```diff
@@ -11,50 +11,65 @@
...
-                const output_terminals_type &outs,  // tuple of pointers to output terminals
-                const std::string &name = "compositett")
-        : TTBase(name, numins, numouts), tts(std::forward<opsT>(ops_take_ownership)), ins(ins), outs(outs) {
-      if (tts.size() == 0) throw name + ":CompositeTT: need to wrap at least one TT";  // see fence
-
-      set_is_composite(true);
-      for (auto &op : tts) op->set_is_within_composite(true, this);
-      set_terminals(ins, &CompositeTT<input_terminalsT, output_terminalsT>::set_input);
-      set_terminals(outs, &CompositeTT<input_terminalsT, output_terminalsT>::set_output);
+    TTG(opsT &&ops,
```
Comment (devreal) [inline ttg/ttg/tt.h:50]: Should this be `ttsT` instead of `opsT`?

File: ttg/ttg/tt.h [hunk 5]
```diff
@@ -65,12 +80,19 @@ namespace ttg {
...
+      for (auto &op : tts) op->owning_ttg = this;
+    }
+
   };
 
   template <typename opsT, typename input_terminalsT, typename output_terminalsT>
-  std::unique_ptr<CompositeTT<input_terminalsT, output_terminalsT>> make_composite_tt(
-      opsT &&ops, const input_terminalsT &ins, const output_terminalsT &outs, const std::string &name = "compositett") {
-    return std::make_unique<CompositeTT<input_terminalsT, output_terminalsT>>(std::forward<opsT>(ops), ins, outs, name);
+  auto make_ttg(opsT &&ops, const input_terminalsT &ins,
```
Comment (devreal) [inline ttg/ttg/tt.h:92]: Same as above: `ttsT` instead of `opsT`?

File: examples/test/test.cc [hunk 6]
```diff
@@ -540,30 +540,32 @@ int try_main(int argc, char **argv) {
...
+      TTG ttg;
+      auto [A, B] = ttg.edges("A", "B");  // TODO: abstract Edge
+      ttg.emplace([&]() {
+        std::cout << "produced 0" << std::endl;
+        sendk(A, 0);
+      })
+      .emplace([&](int key) {
            if (key <= 100) {
-             send<0>(key + 1);
+             sendk(A, key + 1);
```
[the remark this answers is not in the document]
Comment (evaleev) [reply examples/test/test.cc:551]: we can discuss at the ttg meeting

**Pull request merged**